File: Mechanica Engine/src/renderers/Renderers.java
package renderers;

import font.GUIText;
import matrices.TransformationMatrix;
import matrices.ViewMatrix;
import org.lwjgl.util.vector.Matrix4f;
import renderers.color.ColorShader;
import renderers.font.FontShader;
import renderers.texture.TextureShader;
import statics.W;

import static org.lwjgl.opengl.GL11.*;
import static org.lwjgl.opengl.GL13.GL_TEXTURE0;
import static org.lwjgl.opengl.GL13.glActiveTexture;
import static renderers.DefaultRenderer.*;

/**
 * Created by domin on 28 Apr 2017.
 */
public class Renderers {
    public static RenderFunction colorRenderer, textureRenderer, customMatrixRenderer, fontRenderer;
    public static ColorShader colorShader;
    public static TextureShader textureShader;
    public static FontShader fontShader;

    public static float[] colorArray;
    public static TransformationMatrix transformationMatrix;
    public static ViewMatrix viewMatrix;
    public static Matrix4f matrix4f;

    public static void makeShaders(){
        textureShader = new TextureShader();
        textureShader.start();
        textureShader.loadProjectionMatrix(W.projectionMatrix.create());
        textureShader.loadViewMatrix(W.viewMatrix.create());
        textureShader.stop();

        colorShader = new ColorShader();
        colorShader.start();
        colorShader.loadProjectionMatrix(W.projectionMatrix.create());
        colorShader.loadViewMatrix(W.viewMatrix.create());
        colorShader.stop();

        fontShader = new FontShader();
        fontShader.start();
        fontShader.loadProjectionAndView(W.projectionMatrix, W.viewMatrix);
        fontShader.stop();

        colorArray = new float[4];
        transformationMatrix = new TransformationMatrix();
        viewMatrix = W.viewMatrix;
    }

    public static void makeRenderers(){
        colorRenderer = (model) -> {
            prepareVertexArrays(model.getVaoID(), 1);

            colorShader.start();
            colorShader.loadViewMatrix(viewMatrix.create());
            colorShader.loadTransformationMatrix(transformationMatrix.create());
            colorShader.setUniform4fv(colorShader.color, colorArray);

            if (model.drawType== GL_TRIANGLES)
                glDrawElements(GL_TRIANGLES, model.getVertexCount(),
                        GL_UNSIGNED_SHORT, 0);
            else
                glDrawArrays(model.drawType, 0,
                        model.getVertexCount());
            
            colorShader.stop();

            disableVertexArrays();
        };

        textureRenderer = (model) -> {
            prepareVertexArrays(model.getVaoID(), 2);

            textureShader.start();
            textureShader.loadViewMatrix(viewMatrix.create());
            textureShader.loadTransformationMatrix(transformationMatrix.create());

            enableAlphaBlending();
            drawTexture(model);

            disableVertexArrays();
            textureShader.stop();
        };

        fontRenderer = (model) -> {
            prepareVertexArrays(model.getVaoID(), 2);
            enableAlphaBlending();

            GUIText text = (GUIText) model;

            fontShader.start();
            fontShader.loadTranslation(text.getPositionX(), text.getPositionY());
            fontShader.loadColor(colorArray[0], colorArray[1], colorArray[2], colorArray[3]);
            fontShader.loadProjectionAndView(W.projectionMatrix, W.UIView);

            glActiveTexture(GL_TEXTURE0);
            glBindTexture(GL_TEXTURE_2D, model.getTexture());
//		glDrawElements(GL11.GL_TRIANGLES, model.getVertexCount(),
//				GL11.GL_UNSIGNED_SHORT, 0);
            glDrawArrays(GL_TRIANGLES, 0, model.getVertexCount());

            disableVertexArrays();
            fontShader.stop();
        };

        customMatrixRenderer = (model) -> {
            prepareVertexArrays(model.getVaoID(), 2);

            textureShader.start();
            textureShader.loadViewMatrix(viewMatrix.create());
            textureShader.loadTransformationMatrix(matrix4f);

            enableAlphaBlending();
            drawTexture(model);

            disableVertexArrays();
            textureShader.stop();
        };
    }

    public static void cleanUp(){
        textureShader.cleanUp();
        colorShader.cleanUp();
        fontShader.cleanUp();
    }
}

File: lwjgl3-test/src/renderer/Renderers.kt
package renderer

import display.Game
import font.GUIText
import loader.loadTexture
import loader.loadTexturedQuad
import matrices.TransformationMatrix
import models.Model
import org.lwjgl.opengl.GL11
import org.lwjgl.opengl.GL11.*
import org.lwjgl.opengl.GL13
import org.lwjgl.opengl.GL13.GL_TEXTURE0
import org.lwjgl.opengl.GL13.glActiveTexture
import shaders.framebuffer.FXAAShader
import renderer.framebuffer.Fbo
import shaders.circle.CircleShader
import shaders.color.ColorShader
import shaders.font.FontShader
import shaders.texture.TextureShader

/**
 * Created by domin on 27/10/2017.
 */


val transformationMatrix = TransformationMatrix()
var drawingViewMatrix =  Game.viewMatrix

var textureShader = TextureShader()
var textureRenderer:(Model)-> Unit = {
    prepareVertexArrays(it.vaoID, 2)

    textureShader.start()
    textureShader.loadViewMatrix(drawingViewMatrix.create())
    textureShader.loadTransformationMatrix(transformationMatrix.create())

    enableAlphaBlending()
    drawTexture(it)

    disableVertexArrays()
    textureShader.stop()
}

var colorShader = ColorShader()
var colorArray: FloatArray = floatArrayOf(1f,1f,1f,1f)
var colorRenderer:(Model)-> Unit = {
    prepareVertexArrays(it.vaoID, 1)

    colorShader.start()
    colorShader.loadViewMatrix(drawingViewMatrix.create())
    colorShader.loadTransformationMatrix(transformationMatrix.create())
    colorShader.setUniform4fv(colorShader.color, colorArray)

    if (it.drawType == GL_TRIANGLES)
        glDrawElements(GL_TRIANGLES, it.vertexCount,
                GL_UNSIGNED_SHORT, 0)
    else
        glDrawArrays(it.drawType, 0,
                it.vertexCount)

    colorShader.stop()

    disableVertexArrays()
}

val circleShader = CircleShader()
val circleModel = loadTexturedQuad(loadTexture("images/oval"), 0f, 1f, 1f, 0f)
var strokeWidth = 0.1
private var border = 0.05

var circleRenderer:()-> Unit = {
    prepareVertexArrays(circleModel.vaoID, 2)

    circleShader.start()
    circleShader.loadViewMatrix(drawingViewMatrix.create())
    border = 0.03/((transformationMatrix.scaleX + transformationMatrix.scaleY))
    circleShader.loadTransformationMatrix(transformationMatrix.create())
    circleShader.setUniform4fv(circleShader.color, colorArray)
    circleShader.loadFloat(circleShader.strokeWidth, strokeWidth.toFloat())
    circleShader.loadFloat(circleShader.border, border.toFloat())

    enableAlphaBlending()
    drawTexture(circleModel)

    circleShader.stop()

    disableVertexArrays()
}

var fontShader = FontShader()
var fontRenderer:(GUIText)-> Unit = { text ->
    val model = text.model
    prepareVertexArrays(model.vaoID, 2)
    enableAlphaBlending()


    fontShader.start()
    fontShader.loadTranslation(text.positionX, text.positionY)
    fontShader.setUniform4fv(fontShader.color, colorArray)
    fontShader.loadProjectionAndView(Game.projectionMatrix, Game.uiViewMatrix)

    glActiveTexture(GL_TEXTURE0)
    glBindTexture(GL_TEXTURE_2D, model.texture)
//		glDrawElements(GL11.GL_TRIANGLES, model.getVertexCount(),
//				GL11.GL_UNSIGNED_SHORT, 0);
    glDrawArrays(GL_TRIANGLES, 0, model.vertexCount)

    disableVertexArrays()
    fontShader.stop()
}

val inverseRes = floatArrayOf(1f/Game.width.toFloat(), 1f/Game.height.toFloat())
val fbo = Fbo(Game.width, Game.height)
val AAShader = FXAAShader()
var frameRenderer: (Model) -> Unit = {
    prepareVertexArrays(it.vaoID, 1)

    AAShader.start()
    AAShader.setUniform2fv(AAShader.inverseRes, inverseRes)

    GL13.glActiveTexture(GL13.GL_TEXTURE0)
    GL11.glBindTexture(GL11.GL_TEXTURE_2D, it.texture)

    GL11.glDrawElements(GL11.GL_TRIANGLES, it.vertexCount,
            GL11.GL_UNSIGNED_SHORT, 0)

    AAShader.stop()

    disableVertexArrays()
}

val quad: Model = loadTexturedQuad(0, -1f, 1f, 1f, -1f)
var backRenderer = {
    prepareVertexArrays(quad.vaoID, 1)

    glDrawElements(GL_TRIANGLES, quad.vertexCount,
            GL_UNSIGNED_SHORT, 0)

    disableVertexArrays()
}